Make address translation respect the CPU translation mode

magumagu opened the pull request:
The PowerPC CPU has bits in MSR (DR and IR) which control whether addresses are translated.  We should respect these instead of mixing physical addresses and translated addresses into the same address space.

The bulk of this change is mass-renaming: making translation for CPU addresses work correctly means that memory accesses from the CPU have to use a CPU-specific API for memory access.

On its own, this isn't really that useful, because it doesn't change behavior unless a game crashes.  That said, it's the first step to further improvements to the memory subsystem (real BAT emulation, locking, etc.). 

My current naming scheme is to turn Memory::Read/Write* into the hardware variants, and move the CPU versions to PowerPC::Read/Write*. That said, I'm open to suggestions.

~~Fixes issue 8064.~~ This will have to wait for dynamic-bat (#1882).

magumagu commented:
Anyone have comments on the API names?

I think I've settled on the right approach to how the JIT should handle changes to MSR.DR/IR; these can only change on an exception exit, so we basically just need to add a special codepath to exception exits to update the JIT state.  (Not finished/uploaded yet.)  To make this completely correct, the JIT cache actually needs to have separate address spaces for changes to MSR.DR/IR/PR, but I'm going to assume that can be left for a followup because code that tried to abuse that would probably break on master anyway.

skidau commented:
Would you mind rebasing on latest master?

skidau commented:
The API names as you have them are fine by me.  I'd change Debug_Write_U* to the old name of WriteUnchecked_U* or as they are functions that are to be called from host code, what about Write_U*_Host?  TryReadInstruction sounds odd but I don't have a suggested name for that function at this time.

skidau commented:
This PR is as fast if not faster than master.  Games that use the MMU speed hack will need to have the MMU option enabled (or always enable the MMU in this PR).

Super Mario Galaxy hangs in the strap screen.

Toy Story 3 logs "Strange program counter with address translation off" messages but works fine anyway.

Here is the Toy Story 3 log: http://pastie.org/pastes/9824985/text?key=umsc6hkbtuwffxdy3xu14g

JMC47 commented:
Mario Galaxy not working is a bit weird...

magumagu commented:
Pushed a couple of fixes; should help the MMU-hack issue.

I'd appreciate more info on the SMG hang...

Toy Story 3 executing code from the physical address 0x801fb6d4 thing is rather suspicious, but there's a possibility that it's just terrible coding in the game rather than a Dolphin bug.  I would guess that on hardware, the bus would mask off the top two bits and fetch code from 0x001fb6d4, just like it would if the GPU or something tried to access 0x801fb6d4.  I guess I'll change the PanicAlert to an ERROR_LOG; someone should probably verify the hardware behavior at some point.

magumagu commented:
Note to self: accessing physical_base or logical_base outside the JIT is probably a good way to break ARM32.

magumagu commented:
For the names, I picked Debug_Write_U* because the primary use-case is the debugger: we call them if a user uses the memory view to read/modify memory.  (Currently, they're also used by the cheat engine and some other misc stuff, but most of it probably shouldn't be using debug functions.)  I guess Host would work as well.  I'm not a fan of "unchecked" because it describes neither how it is expected to be used nor how the implementation works.

skidau commented:
I'd prefer HostWrite_U* or Write_U*_From_Host over Debug_Write_U* as these functions can be used for things other than debug related.  Personally, I'd overload the functions with an optional parameter at the end and keep the function names uniform.

magumagu commented:
I think this is finished/ready for review.

comex commented on Source/Core/Core/Core.vcxproj.filters:
```diff
@@ -706,6 +703,11 @@
     </ClCompile>
+    <ClCompile Include="HW\SI_GCAdapter.cpp" />
+    <ClCompile Include="PowerPC\JitCommon\JitBackpatch.cpp" />
+    <ClCompile Include="PowerPC\MMU.cpp">
```
What is all this stuff?

comex commented:
Does this ensure that it won't jump to a JIT block which was compiled with different IR/DR?

I can't find such logic but I haven't looked that hard.

magumagu commented:
The code in this PR works properly as long as a game doesn't try to execute code with the same address with varying values of IR/DR without invalidating the JIT cache.  I'm pretty sure games don't do this because master doesn't handle it correctly anyway.  That said, I plan to rewrite the JIT cache in a followup, and I'll make sure I fix that issue then.

All other cases should work fine; the code updates the fastmem base in the block dispatcher, and we never link blocks which end in an instruction which can change DR.

comex commented:
Well, OK, as long as it gets fixed eventually.

skidau commented:
Toy Story 3 stopped working with the Dynamic BAT changes.  The game works with this PR.  It also works in master.

Here's a log of Toy Story 3 booting with the Dynamic BAT changes:
http://pastie.org/pastes/9829408/text?key=73mw24xh4tfipjxdpk1ww

skidau commented:
Another log with the successful mapping logged: http://pastie.org/pastes/9829457/text?key=fy9sc7iakne4buwclplug

magumagu commented:
This is going to get confusing fast if we start talking about that branch here; I'll open a new PR.

magumagu commented:
Current status:

1. This currently causes a minor performance regression; probably not hard to fix.

2. This breaks cheats.  Currently, our cheat code is called once every 60th of a second, completely ignoring CPU state.  This is very messy because the functioning of our cheat handlers depends on CPU state; specifically, trying to run cheats with translation off doesn't work.  Trying to figure out what to do about this; my current thinking is that we should insert a hook, like Gecko OS does, but I'm not sure how to do that in Dolphin.  (Alternatively, I could force on MSR.DR/IR during cheat handling; that seems really evil, but it'll probably make cheats work about as well as they do now.)

JMC47 commented:
Cheats are working in the latest.

Also noticed that the slowdown seems limited to MMU titles... if that makes sense?

magumagu commented:
Yes, that makes sense; only MMU titles use the functions in MMU.cpp enough for performance to matter.

magumagu commented:
Current status: cleaned up a bit more, squashed, fifologs now work.  Still causes slowdown with certain settings on certain scenes in RS3 for reasons which are unclear.  I don't know of anything else that might block merging this.

FioraAeterna commented:
I'll glance through it and see if I have any comments now that it's not WIP anymore

magumagu commented:
It turns out the performance "regression" wasn't really a regression, so there isn't anything blocking this as far as I know.

FioraAeterna commented:
I think this patch looks okay, personally!

FioraAeterna commented:
With making this change, should we drop #1855 and merge it into this? In other words, should host writes to the system be set up to invalidate JIT blocks at those addresses?

magumagu commented:
Hmm... interesting.  They probably should; I can't think of a situation where you'd want to modify memory and not invalidate the jit cache etc.  I'll try it out.

magumagu commented:
As far as I know, there isn't anything blocking merging this.

magumagu commented:
Err, scratch that, apparently ARM64 broke at some point; I'll look into it.

magumagu commented:
Okay, ARM64 should be fixed... might be worth one more round of general testing just to make sure nothing broke since the last round.

Sonicadvance1 commented:
Works fine on AArch64

phire commented:
This is ready for merge, right?

phire commented:
A bit light on the comments, but it looks good to me.

CarlKenner commented:
This is missing a < /ClCompile > !

shuffle2 commented:
@CarlKenner this comment would be more helpful tagged to a file/line...

CarlKenner commented on Source/Core/Core/Core.vcxproj.filters:
```diff
@@ -726,6 +723,8 @@
     </ClCompile>
     <ClCompile Include="HW\GCKeyboard.cpp">
       <Filter>HW %28Flipper/Hollywood%29\GCKeyboard</Filter>
+    <ClCompile Include="PowerPC\MMU.cpp">
```
This is missing a < /ClCompile > !

tueidj commented on Source/Core/Core/HW/Memmap.cpp:
```diff
@@ -102,17 +103,60 @@ bool IsInitialized()
+// Dolphin allocates memory to represent four regions:
+// - 24MB RAM, available on Gamecube and Wii
+// - 64MB "EXRAM", RAM only available on Wii
+// - 64MB FakeVMem, allocated when MMU support is turned off. This is used
```
Why is this 64MB? The "common library" that it refers to is the GC VM SDK, which supports less than 16MB (the physical size of ARAM used for backing) of virtual memory.

magumagu replied:
Err, that's a typo; we actually allocate 32MB.

As to why, I have no idea; it's been done that way since before I ever looked at the code in question.

tueidj commented on Source/Core/Core/PowerPC/MMU.cpp:
```diff
@@ -217,31 +250,71 @@ __forceinline void WriteToHardware(u32 em_address, const T data)
+		return;
+	}
+
+	if (!performTranslation)
```
The PowerPC can't (and hence shouldn't attempt) write to MMIOs with data translation off. The reason being all data access in real mode is performed using WIMG access mode 0011 (write-back, caching enabled) which makes the writes go to the cache instead of directly to the hardware. At some point later the cache line will be evicted, sending 32 bytes across the relevant hardware address range... you can probably guess why this doesn't play nice with real hardware.
Unless there's a good reason to keep this block of code in I would suggest replacing it with an abort() equivalent.

phire replied:
Interesting.

What happens if the caches are disabled or locked?

PsiLupan commented:
This PR is so large that I can't begin to find what caused it, but it broke /debug's ability to view PPC instructions in the Code window. They're still mapped, but everything is <unknown> and only viewable if you use PPC vs x86.

Tested with the release prior to this (4.0-5445) to see that it worked there, and then latest build of Dolphin.

PsiLupan commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -19,26 +19,22 @@
 		return "";
 
-	if (Core::GetState() != Core::CORE_UNINITIALIZED)
+	if (Core::GetState() == Core::CORE_PAUSE)
```
Is there any particular reason this was done?

Iit broke /debug's ability to view PPC instructions in the Code window unless the emulation is paused. They're still mapped, but everything is and only viewable if you use PPC vs x86.

Tested with the release prior to this (4.0-5445) to see that it worked there, and then latest build of Dolphin.

magumagu replied:
The problem is the contents of the code window depends on processor state (whether address translation is turned on, MMU mappings, the contents of memory, etc.).  This state can't be safely read unless emulation is paused; otherwise, you end up with race conditions.  These races became much more evident with this PR because address translation gets turned on and off frequently, leading to PanicAlerts and inconsistent results.

mimimi085181 commented:
@magumagu: This pr broke NeoGamma:
https://bugs.dolphin-emu.org/issues/9470

mimimi085181 commented:
@ magumagu: Did you see that i tracked down where the problem happens in NeoGamma?
https://bugs.dolphin-emu.org/issues/9470

(sorry for abusing this pr, but i don't have any other means to get your attention)

magumagu commented:
@mimimi085181 I'll reply there.  For future reference, my email address is in the git commit log.
